flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp

```diff
@@ -85,67 +135,227 @@ class OMPMapInfoFinalizationPass
+
+    // If the operation that we are expanding with a descriptor has a user
+    // (parent), then we have to expand the parents member indices to reflect
+    // the adjusted member indices for the base address insertion. However, if
+    // it does not then we are expanding a MapInfoOp without any pre-existing
+    // member information to now have one new member for the base address or we
+    // are expanding a parent that is a descriptor and we have to adjust all of
+    // it's members to reflect the insertion of the base address.
+    if (!mapMemberUsers.empty()) {
+      auto baseAddrIndex = memberIndices[mapMemberUsers[0].index];
+      adjustMemberIndices(memberIndices, mapMemberUsers[0].index);
+      llvm::SmallVector<mlir::Value> newMemberOps;
+      mlir::OperandRange membersArr = mapMemberUsers[0].parent.getMembers();
+      for (size_t i = 0; i < membersArr.size(); ++i) {
+        newMemberOps.push_back(membersArr[i]);
+        if (membersArr[i] == op)
+          newMemberOps.push_back(baseAddr);
+      }
+      mapMemberUsers[0].parent.getMembersMutable().assign(newMemberOps);
+      Fortran::lower::omp::fillMemberIndices(memberIndices);
+      mapMemberUsers[0].parent.setMembersIndexAttr(
+          Fortran::lower::omp::createDenseElementsAttrFromIndices(memberIndices,
+                                                                  builder));
+    } else {
+      newMembers.push_back(baseAddr);
+      if (!op.getMembers().empty()) {
+        for (auto &indices : memberIndices)
+          indices.insert(indices.begin(), 0);
+        llvm::SmallVector<int> baseAddrIndex;
+        baseAddrIndex.resize(memberIndices[0].size());
+        std::fill(baseAddrIndex.begin(), baseAddrIndex.end(), -1);
+        baseAddrIndex[0] = 0;
+        memberIndices.insert(memberIndices.begin(), baseAddrIndex);
+        Fortran::lower::omp::fillMemberIndices(memberIndices);
+        newMembersAttr =
+            Fortran::lower::omp::createDenseElementsAttrFromIndices(
+                memberIndices, builder);
+        newMembers.append(op.getMembers().begin(), op.getMembers().end());
+      } else {
+        newMembersAttr = mlir::DenseIntElementsAttr::get(
             mlir::VectorType::get(
                 llvm::ArrayRef<int64_t>({1, 1}),
                 mlir::IntegerType::get(builder.getContext(), 32)),
-            llvm::ArrayRef<int32_t>({0})),
-        /*bounds=*/mlir::SmallVector<mlir::Value>{},
-        builder.getIntegerAttr(builder.getIntegerType(64, false),
-                               op.getMapType().value()),
-        op.getMapCaptureTypeAttr(), op.getNameAttr(), op.getPartialMapAttr());
-    op.replaceAllUsesWith(newDescParentMapOp);
+            llvm::ArrayRef<int32_t>({0}));
+      }
+    }
+
+    mlir::omp::MapInfoOp newDescParentMapOp =
+        builder.create<mlir::omp::MapInfoOp>(
+            op->getLoc(), op.getResult().getType(), descriptor,
+            mlir::TypeAttr::get(fir::unwrapRefType(descriptor.getType())),
```

Comment:
Nit: Document argument names directly initialized in the call, using the `/*name=*/...` format.